plugins/DataStreamCereal/datastream_cereal.cpp

```diff
@@ -0,0 +1,162 @@
+#include <assert.h>
+
+#include <QDebug>
+#include <QDialog>
+#include <QElapsedTimer>
+#include <QIntValidator>
+#include <QMessageBox>
+#include <QSettings>
+
+#include "datastream_cereal.h"
+#include "ui_datastream_cereal.h"
+
+
+StreamCerealDialog::StreamCerealDialog(QWidget *parent) :
+  QDialog(parent),
+  ui(new Ui::DataStreamCereal)
+{
+  ui->setupUi(this);
+}
+
+StreamCerealDialog::~StreamCerealDialog()
+{
+  while (ui->layoutOptions->count() > 0)
+  {
+    auto item = ui->layoutOptions->takeAt(0);
+    item->widget()->setParent(nullptr);
+  }
+  delete ui;
+}
+
+DataStreamCereal::DataStreamCereal():
+  _running(false),
+  show_deprecated(std::getenv("SHOW_DEPRECATED"))
+{
+}
+
+DataStreamCereal::~DataStreamCereal()
+{
+  shutdown();
+}
+
+bool DataStreamCereal::start(QStringList*)
+{
+  if (_running)
+  {
+    return _running;
+  }
+
+  StreamCerealDialog* dialog = new StreamCerealDialog();
+  if (std::getenv("ZMQ"))
+  {
+    qDebug() << "Using ZMQ backend!";
+
+    QSettings settings;
+    address = settings.value("Cereal_Subscriber::address", "192.168.43.1").toString();
+    dialog->ui->lineEditAddress->setText(address);
+
+    int res = dialog->exec();
+    if (res == QDialog::Rejected)
+    {
+      _running = false;
+      return false;
+    }
+    address = dialog->ui->lineEditAddress->text();
+    // save for next time
+    settings.setValue("Cereal_Subscriber::address", address);
+  }
+  else
+  {
+    qDebug() << "Using MSGQ backend!";
+    address = "127.0.0.1";
+  }
+
+  c = Context::create();
+  poller = Poller::create();
+  for (const auto &it : services)
+  {
+    std::string name = std::string(it.name);
+
+    SubSocket *socket;
+    socket = SubSocket::create(c, name, address.toStdString(), false, true);  // don't conflate
+    assert(socket != 0);
+    socket->setTimeout(0);
+
+    poller->registerSocket(socket);
+    sockets.push_back(socket);
+  }
+
+  _running = true;
+  _receive_thread = std::thread(&DataStreamCereal::receiveLoop, this);
+
+  dialog->deleteLater();
+  return _running;
+}
+
+void DataStreamCereal::shutdown()
+{
+  if (_running)
+  {
+    _running = false;
+    if (_receive_thread.joinable())
+    {
+      _receive_thread.join();
+    }
+
+    for (auto sock : sockets)
+    {
+      delete sock;
+    }
+
+    sockets.clear();
```

Comment:
I don't think deleting the items in the vector removes them from the vector. If you don't, you'll run into a segfault from the `delete sock` line above if you stop and restart the plugin a few times. Probably because there's an entrance in the vector for the deleted memory and it tries to delete the already deleted sock.